src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c

```diff
@@ -6,13 +6,256 @@
+    int                     active_mtu;
+    int                     max_mtu;
+    char                    addrstr[128];
+    ucc_tl_mlx5_context_t  *mlx5_ctx;
+    ucc_base_lib_t         *lib;
+
+    ucc_tl_mlx5_mcast_coll_context_t *ctx = NULL;
+
+    mcast_ctx_conf->bcast   = ucc_tl_mlx5_mcast_service_bcast;
+    mcast_ctx_conf->barrier = ucc_tl_mlx5_mcast_service_barrier;
+
+    ctx = &(context->mcast_context);
+    memset(ctx, 0, sizeof(ucc_tl_mlx5_mcast_coll_context_t));
+    memcpy(&ctx->params, mcast_ctx_conf, sizeof(ucc_tl_mlx5_mcast_ctx_params_t));
+
+    mlx5_ctx = ucc_container_of(context, ucc_tl_mlx5_context_t, mcast);
+    lib      = mlx5_ctx->super.super.lib;
+    ctx->lib = lib;
+
+    device_list = ibv_get_device_list(&num_devices);
+    if (!device_list) {
+        tl_error(lib, "TL/MCAST: No IB devices available");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    if (num_devices >= 1 && !strcmp(mcast_ctx_conf->ib_dev_name, "")) {
+        dev          = device_list[0];
+        devname      = (char *)ibv_get_device_name(dev);
+        ctx->devname = ucc_malloc(strlen(devname)+16, "devname");
+        if (!ctx->devname) {
+            status = UCC_ERR_NO_MEMORY;
+            goto error;
+        }
+        strcpy(ctx->devname, devname);
+        strcat(ctx->devname, ":1");
+    } else {
+        ctx->devname = mcast_ctx_conf->ib_dev_name;
+    }
+
+    ibv_free_device_list(device_list);
+    if (UCC_OK != ucc_tl_probe_ip_over_ib(ctx->devname, &ip_oib_addr)) {
+        tl_error(lib, "TL/MCAST: Failed to get ipoib interface for devname %s", ctx->devname);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    is_ipv4     = (ip_oib_addr.ss_family == AF_INET) ? 1 : 0;
+    in_src_addr = (struct sockaddr_in*)&ip_oib_addr;
+
+    inet_ntop((is_ipv4) ? AF_INET : AF_INET6,
+              &in_src_addr->sin_addr, addrstr, sizeof(addrstr) - 1);
+    tl_debug(ctx->lib, "TL/MCAST: devname %s, ipoib %s", ctx->devname, addrstr);
+
+    ctx->channel = rdma_create_event_channel();
+    if (!ctx->channel) {
+        tl_error(lib, "TL/MCAST: rdma_create_event_channel failed, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    memset(&dst_addr, 0, sizeof(struct sockaddr_storage));
+    dst_addr.ss_family = is_ipv4 ? AF_INET : AF_INET6;
+    if (rdma_create_id(ctx->channel, &ctx->id, NULL, RDMA_PS_UDP)) {
+        tl_error(lib, "TL/MCAST: Failed to create rdma id, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    if (0 != rdma_resolve_addr(ctx->id, (struct sockaddr *)&ip_oib_addr,
+                (struct sockaddr *) &dst_addr, 1000)) {
+        tl_error(lib, "TL/MCAST: Failed to resolve rdma addr, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    if (rdma_get_cm_event(ctx->channel, &revent) < 0 ||
+        revent->event != RDMA_CM_EVENT_ADDR_RESOLVED) {
```

Comment:
I might be wrong but in the case where `rdma_get_cm_event` is successfull and
`revent->event != RDMA_CM_EVENT_ADDR_RESOLVED` do we need to free `revent`?